File: src/MrHotel/ApiService/Core/Validation/ValidationResult.cs
namespace MrHotel.ApiService.Core.Validation;

using System.Diagnostics.Contracts;

public sealed class ValidationResult
{
    private readonly IReadOnlyCollection<ValidationError> errors;

    private ValidationResult()
        : this([])
    {
    }

    private ValidationResult(IReadOnlyCollection<ValidationError> errors)
    {
        this.errors = errors;
    }

    public static ValidationResult Success { get; } = new() { Succeeded = true };

    public required bool Succeeded { get; init; }

    public IEnumerable<ValidationError> Errors => this.errors;

    [Pure]
    public static ValidationResult Failed(params ValidationError[] errors)
    {
        ArgumentNullException.ThrowIfNull(errors);

        return new ValidationResult(errors) { Succeeded = false };
    }

    [Pure]
    public static ValidationResult Failed(IEnumerable<ValidationError> errors)
    {
        return Failed(errors.ToArray());
    }

    [Pure]
    public IDictionary<string, string[]> AsErrorDictionary()
    {
        return this.errors.ToDictionary(
            error => error.Name,
            error => new string[] { error.Description });
    }
}

File: src/MrHotel/WebApp/src/app/types/validation/validation-result.ts
import { ValidationError } from "./validation-error";

export class ValidationResult {
  private constructor(
    private readonly _succeeded: boolean = false,
    private readonly _errors: ReadonlyArray<ValidationError>,
  ) {}

  public get succeeded(): boolean {
    return this._succeeded;
  }

  public get errors(): ReadonlyArray<ValidationError> {
    return this._errors;
  }

  public static success() {
    return new ValidationResult(true, []);
  }

  public static failed(...errors: ValidationError[]): ValidationResult {
    if (errors == null) {
      throw new Error("Argument cannot be null or undefined.");
    }

    const failedResult = new ValidationResult(false, errors);
    return failedResult;
  }
}
